Python/Product/PythonTools/PythonTools/Intellisense/InvalidEncodingSquiggleProvider.cs

```diff
@@ -0,0 +1,101 @@
+            if (message != null) {
+                if (!bi.Buffer.Properties.TryGetProperty<string>(VsProjectAnalyzer.InvalidEncodingMoniker, out var prevMessage)
+                    || prevMessage != message) {
+
+                    bi.Buffer.Properties[VsProjectAnalyzer.InvalidEncodingMoniker] = message;
+                    var version = bi.Buffer.CurrentSnapshot.Version;
+
+                    TaskProvider.ReplaceItems(
+                        bi.Filename,
+                        VsProjectAnalyzer.InvalidEncodingMoniker,
+                        new List<TaskProviderItem> {
+                            new TaskProviderItem(
+                                Services.Site,
+                                message,
+                                new SnapshotSpan(bi.CurrentSnapshot, magicEncodingIndex, magicEncodingName.Length).ToSourceSpan(),
+                                VSTASKPRIORITY.TP_NORMAL,
+                                VSTASKCATEGORY.CAT_CODESENSE,
+                                true,
+                                new LocationTracker(version, bi.Buffer, version.VersionNumber)
+                            )
+                        });
+                }
+            } else {
+                TaskProvider.Clear(bi.Filename, VsProjectAnalyzer.InvalidEncodingMoniker);
+                bi.Buffer.Properties.RemoveProperty(VsProjectAnalyzer.InvalidEncodingMoniker);
+            }
+        }
+
+        internal static string CheckEncoding(ITextSnapshot snapshot, Encoding documentEncoding, out string magicEncodingName, out int magicEncodingIndex) {
+            var chunk = snapshot.GetText(new Span(0, Math.Min(snapshot.Length, 512)));
+            Parser.GetEncodingFromMagicDesignator(chunk, out var encoding, out magicEncodingName, out magicEncodingIndex);
+
+            string message = null;
+            if (encoding != null) {
+                // Encoding is specified and is a valid name. 
+                // Check if it matches encoding set on the document text buffer. 
+                if (encoding.EncodingName != documentEncoding.EncodingName) {
+                    message = string.Format(CultureInfo.InvariantCulture, Strings.WarningEncodingMismatch, documentEncoding.EncodingName, magicEncodingName);
+                }
+            } else {
+                if (!string.IsNullOrEmpty(magicEncodingName)) {
+                    // Encoding is specified but not recognized as a valid name
+                    message = string.Format(CultureInfo.InvariantCulture, Strings.WarningInvalidEncoding, magicEncodingName);
+                } else {
+                    // Encoding is not specified. Python assumes UTF-8 so we need to verify it.
+                    if (Encoding.UTF8.EncodingName != documentEncoding.EncodingName) {
+                        message = string.Format(CultureInfo.InvariantCulture, Strings.WarningEncodingDifferentFromDefault, documentEncoding.EncodingName);
```

Comment:
We have helper extension methods on string for `.FormatInvariant()` and `.FormatUI()`. You actually want `.FormatUI()` here (not that it really matters given it's just string concatenation and we have no RTL).